## - Creating <font color='blue'>m18_Global_QLQ30groups65gl</font> feature based on M18_Global_QLQ30
## - Using features which affect QLQ-C30 from <u>bibliography/research</u> ,  <font color='blue'>m0</font> and <font color='blue'>m3</font>
## - SVM - Support Vector Machine
## - SMOTE and without SMOTE 

In [40]:
#importing libraries
import numpy as np
import matplotlib.pyplot as plt
import pandas as pd
import seaborn as sns
import pyreadstat

In [41]:
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

In [42]:
cols = ['age','cancer_stage','M0_performancestatus','M3_performancestatus','income_3','M0_optimism_LOT','radiotherapy','endocrine','antiher2','chemotherapy','care_team','education_3','education_2','M0_employment_status','M3_employment_status','marital_status','marital_status_2','cancer_grade','M0_surgery','M3_surgery','M3_emotional_support','M3_social_support_total','M0_employment_status','income_3','M0_mental_health_total_HADS','M3_mental_health_total_HADS','nccn_distress_thermometer','m3_nccn_distress_thermometer','M0_exercise','M3_PTGI_spiritual_change','baseline_histological_type','baseline_chronic_illness','m0_bmi','m0_Global_QLQ30groups65gl','m3_Global_QLQ30groups65gl','M18_Global_QLQ30']
df=pd.read_spss(r"C:\Users\Chris\Desktop\Cancer_Prediction\BOUNCE_Full Dataset_m0_m3_m6_m9_m12_m15_m18_09-12-2021_ORIGINAL.sav", usecols=cols)

In [43]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status  income_3 M0_exercise  nccn_distress_thermometer  \
0     employed full time  very low    moderate                        0.0   
1     employed full time   average          no                        4.0   
2          self-employed  very low          no                        3.0   
3              housewife  very low          no                        0.0   
4              housewife      high    moderate                        0.5   
..                   ...       ...         ...                        ...   
761                  NaN       NaN         NaN                        NaN   
762        self-employed   average    moderate                        6.0   
763                  NaN       NaN         NaN                        5.2   
764                  NaN       NaN         NaN                        NaN   
765                  NaN       NaN         NaN                        NaN   

     ...  baseline_histological_type  M0_performancestatus  chemotherapy  \
0    ...                      Ductal                   0.0            No   
1    ...                      Ductal                   1.0           Yes   
2    ...                      Ductal                   0.0           Yes   
3    ...                      Ductal                   0.0            No   
4    ...                      Ductal                   0.0            No   
..   ...                         ...                   ...           ...   
761  ...                     Lobular                   0.0            No   
762  ...                      Ductal                   0.0           Yes   
763  ...                      Ductal                   0.0           Yes   
764  ...                         NaN                   NaN           NaN   
765  ...                         NaN                   NaN           NaN   

     radiotherapy  endocrine  M0_surgery  M3_surgery  \
0             Yes        Yes         1.0         0.0   
1             Yes        Yes         0.0         0.0   
2             Yes         No         0.0         NaN   
3             Yes        Yes         0.0         0.0   
4             Yes        Yes         0.0         0.0   
..            ...        ...         ...         ...   
761           Yes        Yes         NaN         0.0   
762            No        Yes         0.0         0.0   
763           Yes        Yes  

In [44]:
# create a list of our conditions (breakpoint 65)
# >65  ->  0
# <65  ->  1
conditions = [(df['M18_Global_QLQ30'] >= 0) & (df['M18_Global_QLQ30'] <= 65),(df['M18_Global_QLQ30'] > 65) & (df['M18_Global_QLQ30'] <= 100)]

In [45]:
# create a list of the values we want to assign for each condition
values = [9, 8]

In [46]:
df['m18_Global_QLQ30groups65gl'] = np.select(conditions, values)

In [47]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status  income_3 M0_exercise  nccn_distress_thermometer  \
0     employed full time  very low    moderate                        0.0   
1     employed full time   average          no                        4.0   
2          self-employed  very low          no                        3.0   
3              housewife  very low          no                        0.0   
4              housewife      high    moderate                        0.5   
..                   ...       ...         ...                        ...   
761                  NaN       NaN         NaN                        NaN   
762        self-employed   average    moderate                        6.0   
763                  NaN       NaN         NaN                        5.2   
764                  NaN       NaN         NaN                        NaN   
765                  NaN       NaN         NaN                        NaN   

     ...  M0_performancestatus  chemotherapy  radiotherapy  endocrine  \
0    ...                   0.0            No           Yes        Yes   
1    ...                   1.0           Yes           Yes        Yes   
2    ...                   0.0           Yes           Yes         No   
3    ...                   0.0            No           Yes        Yes   
4    ...                   0.0            No           Yes        Yes   
..   ...                   ...           ...           ...        ...   
761  ...                   0.0            No           Yes        Yes   
762  ...                   0.0           Yes            No        Yes   
763  ...                   0.0           Yes           Yes        Yes   
764  ...                   NaN           NaN           NaN        NaN   
765  ...                   NaN           NaN           NaN        NaN   

     M0_surgery  M3_surgery  m0_Global_QLQ30groups65gl  \
0           1.0         0.0                   good QoL   
1           0.0         0.0                   good QoL   
2           0.0         NaN                   good QoL   
3           0.0         0.0                   good QoL   
4           0.0         0.0                   good QoL   
..          ...         ...                        ...   
761         NaN         0.0                        NaN   
762         0.0         0.0                   good QoL   
763         0.0         0.0                     

In [48]:
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(0, np.NaN)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(9, 1)
df['m18_Global_QLQ30groups65gl']=df['m18_Global_QLQ30groups65gl'].replace(8, 0)

In [49]:
df = df.drop('M18_Global_QLQ30', 1)

C:\Users\Chris\AppData\Local\Temp/ipykernel_19056/2872487555.py:1: FutureWarning: In a future version of pandas all arguments of DataFrame.drop except for the argument 'labels' will be keyword-only
  df = df.drop('M18_Global_QLQ30', 1)


In [50]:
df

age care_team  education_3                        education_2  \
0    66.0     CHAMP          1.0                          0-9 years   
1    54.0     CHAMP          2.0  high-school/vocational/University   
2    45.0     CHAMP          2.0  high-school/vocational/University   
3    70.0     CHAMP          1.0                          0-9 years   
4    49.0     CHAMP          3.0  high-school/vocational/University   
..    ...       ...          ...                                ...   
761   NaN       IEO          NaN                                NaN   
762  43.0       IEO          3.0  high-school/vocational/University   
763   NaN       IEO          NaN                                NaN   
764   NaN     CHAMP          NaN                                NaN   
765   NaN      HUJI          NaN                                NaN   

         marital_status                      marital_status_2  \
0               Widowed                Single/widowed/engaged   
1               Widowed                Single/widowed/engaged   
2    Common law partner  married/separated/common law partner   
3               Widowed                Single/widowed/engaged   
4               Married  married/separated/common law partner   
..                  ...                                   ...   
761                 NaN                                   NaN   
762  Common law partner  married/separated/common law partner   
763                 NaN                                   NaN   
764                 NaN                                   NaN   
765                 NaN                                   NaN   

    M0_employment_status  income_3 M0_exercise  nccn_distress_thermometer  \
0     employed full time  very low    moderate                        0.0   
1     employed full time   average          no                        4.0   
2          self-employed  very low          no                        3.0   
3              housewife  very low          no                        0.0   
4              housewife      high    moderate                        0.5   
..                   ...       ...         ...                        ...   
761                  NaN       NaN         NaN                        NaN   
762        self-employed   average    moderate                        6.0   
763                  NaN       NaN         NaN                        5.2   
764                  NaN       NaN         NaN                        NaN   
765                  NaN       NaN         NaN                        NaN   

     ...  baseline_histological_type  M0_performancestatus  chemotherapy  \
0    ...                      Ductal                   0.0            No   
1    ...                      Ductal                   1.0           Yes   
2    ...                      Ductal                   0.0           Yes   
3    ...                      Ductal                   0.0            No   
4    ...                      Ductal                   0.0            No   
..   ...                         ...                   ...           ...   
761  ...                     Lobular                   0.0            No   
762  ...                      Ductal                   0.0           Yes   
763  ...                      Ductal                   0.0           Yes   
764  ...                         NaN                   NaN           NaN   
765  ...                         NaN                   NaN           NaN   

     radiotherapy  endocrine  M0_surgery  M3_surgery  \
0             Yes        Yes         1.0         0.0   
1             Yes        Yes         0.0         0.0   
2             Yes         No         0.0         NaN   
3             Yes        Yes         0.0         0.0   
4             Yes        Yes         0.0         0.0   
..            ...        ...         ...         ...   
761           Yes        Yes         NaN         0.0   
762            No        Yes         0.0         0.0   
763           Yes        Yes  

<b>0</b> -> <font color='blue'>good QoL</font> 
<br>
<b>1</b> -> <font color='red'>poor QoL</font> 

In [51]:
#check the category features
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 30 columns):
 #   Column                        Non-Null Count  Dtype   
---  ------                        --------------  -----   
 0   age                           757 non-null    float64 
 1   care_team                     766 non-null    category
 2   education_3                   750 non-null    float64 
 3   education_2                   750 non-null    category
 4   marital_status                745 non-null    category
 5   marital_status_2              745 non-null    category
 6   M0_employment_status          756 non-null    category
 7   income_3                      710 non-null    category
 8   M0_exercise                   749 non-null    category
 9   nccn_distress_thermometer     595 non-null    float64 
 10  m3_nccn_distress_thermometer  617 non-null    float64 
 11  M0_optimism_LOT               724 non-null    float64 
 12  M0_mental_health_total_HADS   745 non-null    floa

In [52]:
#convert category to numeric value
from sklearn.preprocessing import LabelEncoder
labelencoder_Y = LabelEncoder()
df.iloc[:,1] = labelencoder_Y.fit_transform(df.iloc[:,1].values)
df.iloc[:,3] = labelencoder_Y.fit_transform(df.iloc[:,3].values)
df.iloc[:,4] = labelencoder_Y.fit_transform(df.iloc[:,4].values)
df.iloc[:,5] = labelencoder_Y.fit_transform(df.iloc[:,5].values)
df.iloc[:,6] = labelencoder_Y.fit_transform(df.iloc[:,6].values)
df.iloc[:,7] = labelencoder_Y.fit_transform(df.iloc[:,7].values)
df.iloc[:,8] = labelencoder_Y.fit_transform(df.iloc[:,8].values)
df.iloc[:,19] = labelencoder_Y.fit_transform(df.iloc[:,19].values)
df.iloc[:,20] = labelencoder_Y.fit_transform(df.iloc[:,20].values)
df.iloc[:,22] = labelencoder_Y.fit_transform(df.iloc[:,22].values)
df.iloc[:,23] = labelencoder_Y.fit_transform(df.iloc[:,23].values)
df.iloc[:,24] = labelencoder_Y.fit_transform(df.iloc[:,24].values)
df.iloc[:,27] = labelencoder_Y.fit_transform(df.iloc[:,27].values)
df.iloc[:,28] = labelencoder_Y.fit_transform(df.iloc[:,28].values)

In [53]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0          0          1.0            0               4   
1    54.0          0          2.0            1               4   
2    45.0          0          2.0            1               0   
3    70.0          0          1.0            0               4   
4    49.0          0          3.0            1               1   
..    ...        ...          ...          ...             ...   
761   NaN          3          NaN            2               5   
762  43.0          3          3.0            1               0   
763   NaN          3          NaN            2               5   
764   NaN          0          NaN            2               5   
765   NaN          1          NaN            2               5   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                   0                     0         2            1   
1                   0                     0         0            2   
2                   1                     4         2            2   
3                   0                     2         2            2   
4                   1                     2         1            1   
..                ...                   ...       ...          ...   
761                 2                     6         3            3   
762                 1                     4         0            1   
763                 2                     6         3            3   
764                 2                     6         3            3   
765                 2                     6         3            3   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                           0   
1                          4.0  ...                           0   
2                          3.0  ...                           0   
3                          0.0  ...                           0   
4                          0.5  ...                           0   
..                         ...  ...                         ...   
761                        NaN  ...                           1   
762                        6.0  ...                           0   
763                        5.2  ...                           0   
764                        NaN  ...                           3   
765                        NaN  ...                           3   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0             0             1          1         1.0   
1                     1.0             1             1          1         0.0   
2                     0.0             1             1          0         0.0   
3                     0.0             0             1          1         0.0   
4                     0.0             0             1          1         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0             0             1          1         NaN   
762                   0.0             1             0          1         0.0   
763                   0.0             1             1          1         0.0   
764                   NaN             2             2          2         NaN   
765                   NaN             2             2          2         NaN   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                          0                          0   
1           0.0                          0                          0   
2           NaN                          0                          2   
3           0.0                          0                          0   
4           0.0                          0                          0   
..          ...                        ...                        ...   
761         0.0                          2                          0   
762         0.0       

In [54]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 30 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           757 non-null    float64
 1   care_team                     766 non-null    int32  
 2   education_3                   750 non-null    float64
 3   education_2                   766 non-null    int32  
 4   marital_status                766 non-null    int32  
 5   marital_status_2              766 non-null    int32  
 6   M0_employment_status          766 non-null    int32  
 7   income_3                      766 non-null    int32  
 8   M0_exercise                   766 non-null    int32  
 9   nccn_distress_thermometer     595 non-null    float64
 10  m3_nccn_distress_thermometer  617 non-null    float64
 11  M0_optimism_LOT               724 non-null    float64
 12  M0_mental_health_total_HADS   745 non-null    float64
 13  M3_em

In [55]:
#convert corresponding int32 feature type to float64
df['care_team'] = df.care_team.astype(float)
df['education_2'] = df.education_2.astype(float)
df['marital_status'] = df.marital_status.astype(float)
df['marital_status_2'] = df.marital_status_2.astype(float)
df['M0_employment_status'] = df.M0_employment_status.astype(float)
df['income_3'] = df.income_3.astype(float)
df['M0_exercise'] = df.M0_exercise.astype(float)
df['baseline_chronic_illness'] = df.baseline_chronic_illness.astype(float)
df['baseline_histological_type'] = df.baseline_histological_type.astype(float)
df['chemotherapy'] = df.chemotherapy.astype(float)
df['radiotherapy'] = df.radiotherapy.astype(float)
df['endocrine'] = df.endocrine.astype(float)
df['m0_Global_QLQ30groups65gl'] = df.m0_Global_QLQ30groups65gl.astype(float)
df['m3_Global_QLQ30groups65gl'] = df.m3_Global_QLQ30groups65gl.astype(float)

In [56]:
df.info(verbose=True)

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 766 entries, 0 to 765
Data columns (total 30 columns):
 #   Column                        Non-Null Count  Dtype  
---  ------                        --------------  -----  
 0   age                           757 non-null    float64
 1   care_team                     766 non-null    float64
 2   education_3                   750 non-null    float64
 3   education_2                   766 non-null    float64
 4   marital_status                766 non-null    float64
 5   marital_status_2              766 non-null    float64
 6   M0_employment_status          766 non-null    float64
 7   income_3                      766 non-null    float64
 8   M0_exercise                   766 non-null    float64
 9   nccn_distress_thermometer     595 non-null    float64
 10  m3_nccn_distress_thermometer  617 non-null    float64
 11  M0_optimism_LOT               724 non-null    float64
 12  M0_mental_health_total_HADS   745 non-null    float64
 13  M3_em

In [57]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          2.0             5.0   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          2.0             5.0   
764   NaN        0.0          NaN          2.0             5.0   
765   NaN        1.0          NaN          2.0             5.0   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                 0.0                   0.0       2.0          1.0   
1                 0.0                   0.0       0.0          2.0   
2                 1.0                   4.0       2.0          2.0   
3                 0.0                   2.0       2.0          2.0   
4                 1.0                   2.0       1.0          1.0   
..                ...                   ...       ...          ...   
761               2.0                   6.0       3.0          3.0   
762               1.0                   4.0       0.0          1.0   
763               2.0                   6.0       3.0          3.0   
764               2.0                   6.0       3.0          3.0   
765               2.0                   6.0       3.0          3.0   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                         0.0   
1                          4.0  ...                         0.0   
2                          3.0  ...                         0.0   
3                          0.0  ...                         0.0   
4                          0.5  ...                         0.0   
..                         ...  ...                         ...   
761                        NaN  ...                         1.0   
762                        6.0  ...                         0.0   
763                        5.2  ...                         0.0   
764                        NaN  ...                         3.0   
765                        NaN  ...                         3.0   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0           0.0           1.0        1.0         1.0   
1                     1.0           1.0           1.0        1.0         0.0   
2                     0.0           1.0           1.0        0.0         0.0   
3                     0.0           0.0           1.0        1.0         0.0   
4                     0.0           0.0           1.0        1.0         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0           0.0           1.0        1.0         NaN   
762                   0.0           1.0           0.0        1.0         0.0   
763                   0.0           1.0           1.0        1.0         0.0   
764                   NaN           2.0           2.0        2.0         NaN   
765                   NaN           2.0           2.0        2.0         NaN   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                        0.0                        0.0   
1           0.0                        0.0                        0.0   
2           NaN                        0.0                        2.0   
3           0.0                        0.0                        0.0   
4           0.0                        0.0                        0.0   
..          ...                        ...                        ...   
761         0.0                        2.0                        0.0   
762         0.0       

In [58]:
#convert corresponding numeric value to NaN
df['education_2']=df['education_2'].replace(2, np.NaN)
df['marital_status']=df['marital_status'].replace(5, np.NaN)
df['marital_status_2']=df['marital_status_2'].replace(2, np.NaN)
df['M0_employment_status']=df['M0_employment_status'].replace(6, np.NaN)
df['income_3']=df['income_3'].replace(3, np.NaN)
df['M0_exercise']=df['M0_exercise'].replace(3, np.NaN)
df['baseline_chronic_illness']=df['baseline_chronic_illness'].replace(2, np.NaN)
df['baseline_histological_type']=df['baseline_histological_type'].replace(3, np.NaN)
df['chemotherapy']=df['chemotherapy'].replace(2, np.NaN)
df['radiotherapy']=df['radiotherapy'].replace(2, np.NaN)
df['endocrine']=df['endocrine'].replace(2, np.NaN)
df['m0_Global_QLQ30groups65gl']=df['m0_Global_QLQ30groups65gl'].replace(2, np.NaN)
df['m3_Global_QLQ30groups65gl']=df['m3_Global_QLQ30groups65gl'].replace(2, np.NaN)

In [59]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          NaN             NaN   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          NaN             NaN   
764   NaN        0.0          NaN          NaN             NaN   
765   NaN        1.0          NaN          NaN             NaN   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                 0.0                   0.0       2.0          1.0   
1                 0.0                   0.0       0.0          2.0   
2                 1.0                   4.0       2.0          2.0   
3                 0.0                   2.0       2.0          2.0   
4                 1.0                   2.0       1.0          1.0   
..                ...                   ...       ...          ...   
761               NaN                   NaN       NaN          NaN   
762               1.0                   4.0       0.0          1.0   
763               NaN                   NaN       NaN          NaN   
764               NaN                   NaN       NaN          NaN   
765               NaN                   NaN       NaN          NaN   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                         0.0   
1                          4.0  ...                         0.0   
2                          3.0  ...                         0.0   
3                          0.0  ...                         0.0   
4                          0.5  ...                         0.0   
..                         ...  ...                         ...   
761                        NaN  ...                         1.0   
762                        6.0  ...                         0.0   
763                        5.2  ...                         0.0   
764                        NaN  ...                         NaN   
765                        NaN  ...                         NaN   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0           0.0           1.0        1.0         1.0   
1                     1.0           1.0           1.0        1.0         0.0   
2                     0.0           1.0           1.0        0.0         0.0   
3                     0.0           0.0           1.0        1.0         0.0   
4                     0.0           0.0           1.0        1.0         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0           0.0           1.0        1.0         NaN   
762                   0.0           1.0           0.0        1.0         0.0   
763                   0.0           1.0           1.0        1.0         0.0   
764                   NaN           NaN           NaN        NaN         NaN   
765                   NaN           NaN           NaN        NaN         NaN   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                        0.0                        0.0   
1           0.0                        0.0                        0.0   
2           NaN                        0.0                        NaN   
3           0.0                        0.0                        0.0   
4           0.0                        0.0                        0.0   
..          ...                        ...                        ...   
761         0.0                        NaN                        0.0   
762         0.0       

------------------

<font color='blue'><b>MICE Data Imoutation</b></font>

In [21]:
imputer = IterativeImputer(imputation_order='ascending',max_iter=10,random_state=42,n_nearest_features=None)

In [22]:
df = imputer.fit_transform(df)

c:\users\chris\appdata\local\programs\python\python39\lib\site-packages\sklearn\impute\_iterative.py:699: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(


In [23]:
df

array([[66.        ,  0.        ,  1.        , ...,  0.        ,
         0.        ,  0.        ],
       [54.        ,  0.        ,  2.        , ...,  0.        ,
         0.        ,  0.        ],
       [45.        ,  0.        ,  2.        , ...,  0.        ,
         0.32357608,  0.7108489 ],
       ...,
       [49.2317621 ,  3.        ,  2.41397559, ...,  0.39633084,
         0.        ,  0.23687789],
       [53.57624051,  0.        ,  2.65245026, ...,  0.19735294,
         0.36529   ,  0.        ],
       [55.24141938,  1.        ,  2.40575474, ...,  0.35489805,
         0.42146462,  1.        ]])

In [24]:
#Convert Array to Dataframe and name columns
df=pd.DataFrame(df,columns = ['age','care_team','education_3','education_2','marital_status','marital_status_2','M0_employment_status','income_3','M0_exercise','nccn_distress_thermometer','m3_nccn_distress_thermometer','M0_optimism_LOT','M0_mental_health_total_HADS','M3_emotional_support','M3_social_support_total','M3_PTGI_spiritual_change','M3_mental_health_total_HADS','cancer_stage','cancer_grade','baseline_chronic_illness','baseline_histological_type','M0_performancestatus','chemotherapy','radiotherapy','endocrine','M0_surgery','M3_surgery','m0_Global_QLQ30groups65gl','m3_Global_QLQ30groups65gl','m18_Global_QLQ30groups65gl'])

In [25]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.000000        0.0     1.000000     0.000000        4.000000   
1    54.000000        0.0     2.000000     1.000000        4.000000   
2    45.000000        0.0     2.000000     1.000000        0.000000   
3    70.000000        0.0     1.000000     0.000000        4.000000   
4    49.000000        0.0     3.000000     1.000000        1.000000   
..         ...        ...          ...          ...             ...   
761  53.987164        3.0     2.338677     0.890654        1.201495   
762  43.000000        3.0     3.000000     1.000000        0.000000   
763  49.231762        3.0     2.413976     0.903117        1.325372   
764  53.576241        0.0     2.652450     0.939214        1.511978   
765  55.241419        1.0     2.405755     0.922004        1.161582   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0            0.000000              0.000000  2.000000     1.000000   
1            0.000000              0.000000  0.000000     2.000000   
2            1.000000              4.000000  2.000000     2.000000   
3            0.000000              2.000000  2.000000     2.000000   
4            1.000000              2.000000  1.000000     1.000000   
..                ...                   ...       ...          ...   
761          0.914681              1.420556  0.705000     1.033242   
762          1.000000              4.000000  0.000000     1.000000   
763          0.837788              1.132086  0.578031     1.047081   
764          0.817116              0.976014  0.532641     0.952106   
765          0.905185              1.493330  0.792301     1.196854   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                     0.000000  ...                    0.000000   
1                     4.000000  ...                    0.000000   
2                     3.000000  ...                    0.000000   
3                     0.000000  ...                    0.000000   
4                     0.500000  ...                    0.000000   
..                         ...  ...                         ...   
761                   3.949962  ...                    1.000000   
762                   6.000000  ...                    0.000000   
763                   5.200000  ...                    0.000000   
764                   3.251755  ...                    0.297261   
765                   4.619495  ...                    0.296737   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                0.000000      0.000000      1.000000   1.000000    1.000000   
1                1.000000      1.000000      1.000000   1.000000    0.000000   
2                0.000000      1.000000      1.000000   0.000000    0.000000   
3                0.000000      0.000000      1.000000   1.000000    0.000000   
4                0.000000      0.000000      1.000000   1.000000    0.000000   
..                    ...           ...           ...        ...         ...   
761              0.000000      0.000000      1.000000   1.000000    0.294503   
762              0.000000      1.000000      0.000000   1.000000    0.000000   
763              0.000000      1.000000      1.000000   1.000000    0.000000   
764              0.083082      0.629235      0.962287   0.847046    0.272645   
765              0.087655      0.484989      0.726086   0.771273    0.188706   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0      0.000000                   0.000000                   0.000000   
1      0.000000                   0.000000                   0.000000   
2      0.012738                   0.000000                   0.323576   
3      0.000000                   0.000000                   0.000000   
4      0.000000                   0.000000                   0.000000   
..          ...                        ...                        ...   
761    0.000000                   0.1665

<b>- roundup values</b> (<i>age</i>, <i>education_3</i>, <i>education_2</i>, <i>marital_status</i>,<i>marital_status_2</i>, <i>M0_employment_status</i>, <i>income_3</i>, <i>M0_exercise</i>,<i>nccn_distress_thermometer</i>, <i>m3_nccn_distress_thermometer</i>, <i>M0_optimism_LOT</i>, <i>M0_mental_health_total_HADS</i>,<i>M3_emotional_support</i>, <i>M3_social_support_total</i>, <i>M3_PTGI_spiritual_change</i>, <i>M3_mental_health_total_HADS</i>,<i>cancer_stage</i>, <i>cancer_grade</i>, <i>baseline_chronic_illness</i>, <i>baseline_histological_type</i>,<i>M0_performancestatus</i>, <i>chemotherapy</i>,<i>radiotherapy</i>,<i>endocrine</i>, <i>M0_surgery</i>,<i>M3_surgery</i>,<i>m0_Global_QLQ30groups65gl</i>,<i>m3_Global_QLQ30groups65gl</i>, <i>m18_Global_QLQ30groups65gl</i>) -

In [26]:
df.age = df.age.round()
df.education_3 = df.education_3.round()
df.education_2 = df.education_2.round()
df.marital_status = df.marital_status.round()
df.marital_status_2 = df.marital_status_2.round()
df.M0_employment_status = df.M0_employment_status.round()
df.income_3 = df.income_3.round()
df.M0_exercise = df.M0_exercise.round()
df.nccn_distress_thermometer = df.nccn_distress_thermometer.round()
df.m3_nccn_distress_thermometer = df.m3_nccn_distress_thermometer.round()
df.M0_optimism_LOT = df.M0_optimism_LOT.round()
df.M0_mental_health_total_HADS = df.M0_mental_health_total_HADS.round()
df.M3_emotional_support = df.M3_emotional_support.round()
df.M3_social_support_total = df.M3_social_support_total.round()
df.M3_PTGI_spiritual_change = df.M3_PTGI_spiritual_change.round()
df.M3_mental_health_total_HADS = df.M3_mental_health_total_HADS.round()
df.cancer_stage = df.cancer_stage.round()
df.baseline_chronic_illness = df.baseline_chronic_illness.round()
df.baseline_histological_type = df.baseline_histological_type.round()
df.M0_performancestatus = df.M0_performancestatus.round()
df.chemotherapy = df.chemotherapy.round()
df.radiotherapy = df.radiotherapy.round()
df.endocrine = df.endocrine.round()
df.M0_surgery = df.M0_surgery.round()
df.M3_surgery = df.M3_surgery.round()
df.m0_Global_QLQ30groups65gl = df.m0_Global_QLQ30groups65gl.round()
df.m3_Global_QLQ30groups65gl = df.m3_Global_QLQ30groups65gl.round()
df.m18_Global_QLQ30groups65gl = df.m18_Global_QLQ30groups65gl.round()

In [27]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761  54.0        3.0          2.0          1.0             1.0   
762  43.0        3.0          3.0          1.0             0.0   
763  49.0        3.0          2.0          1.0             1.0   
764  54.0        0.0          3.0          1.0             2.0   
765  55.0        1.0          2.0          1.0             1.0   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                 0.0                   0.0       2.0          1.0   
1                 0.0                   0.0       0.0          2.0   
2                 1.0                   4.0       2.0          2.0   
3                 0.0                   2.0       2.0          2.0   
4                 1.0                   2.0       1.0          1.0   
..                ...                   ...       ...          ...   
761               1.0                   1.0       1.0          1.0   
762               1.0                   4.0       0.0          1.0   
763               1.0                   1.0       1.0          1.0   
764               1.0                   1.0       1.0          1.0   
765               1.0                   1.0       1.0          1.0   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                         0.0   
1                          4.0  ...                         0.0   
2                          3.0  ...                         0.0   
3                          0.0  ...                         0.0   
4                          0.0  ...                         0.0   
..                         ...  ...                         ...   
761                        4.0  ...                         1.0   
762                        6.0  ...                         0.0   
763                        5.0  ...                         0.0   
764                        3.0  ...                         0.0   
765                        5.0  ...                         0.0   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0           0.0           1.0        1.0         1.0   
1                     1.0           1.0           1.0        1.0         0.0   
2                     0.0           1.0           1.0        0.0         0.0   
3                     0.0           0.0           1.0        1.0         0.0   
4                     0.0           0.0           1.0        1.0         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0           0.0           1.0        1.0         0.0   
762                   0.0           1.0           0.0        1.0         0.0   
763                   0.0           1.0           1.0        1.0         0.0   
764                   0.0           1.0           1.0        1.0         0.0   
765                   0.0           0.0           1.0        1.0         0.0   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                        0.0                        0.0   
1           0.0                        0.0                        0.0   
2           0.0                        0.0                        0.0   
3           0.0                        0.0                        0.0   
4           0.0                        0.0                        0.0   
..          ...                        ...                        ...   
761         0.0                        0.0                        0.0   
762         0.0       

----------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [28]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [29]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df.iloc[:,1:-1].values
Y=df.iloc[:,-1].values

In [30]:
classifier = SVC(kernel='rbf')

In [31]:
scores = []
cv = KFold(n_splits=10, random_state=42, shuffle=True)
for train_index, test_index in cv.split(X):
    X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]
    classifier.fit(X_train, Y_train)
    scores.append(classifier.score(X_test, Y_test))
ac_svm=np.mean(cross_val_score(classifier, X, Y, cv=10))
print(np.mean(cross_val_score(classifier, X, Y, cv=10)))

0.7612269309637731


<font color='red'>Accuracy: 76.12%</font>

In [32]:
y_predict = classifier.predict(X_test)

In [33]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0  1.0
row_0          
0.0     47    6
1.0     12   11

For 0: <font color='blue'>89%</font> (47/53) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>48%</font> (11/23) correctly classified -> <font color='red'>poor QoL</font> accuracy

------------

<u><font color='blue'>SMOTE</u></font>

In [34]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [35]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [36]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 447, 1.0: 243})
After SMOTE:  Counter({0.0: 447, 1.0: 447})


In [37]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.8157894736842105


<font color='red'>Accuracy: 81,58%</font>

In [39]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     45    8
1.0      6   17

For 0: <font color='blue'>85%</font> (45/53) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>74%</font> (17/23) correctly classified -> <font color='red'>poor QoL</font> accuracy

--------------

<font color='blue'><b><u>without</b></u></font> MICE Data Imoutation <br>
replacing remaining NaN values with the corresponding <u>mean value</u>

In [60]:
df

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761   NaN        3.0          NaN          NaN             NaN   
762  43.0        3.0          3.0          1.0             0.0   
763   NaN        3.0          NaN          NaN             NaN   
764   NaN        0.0          NaN          NaN             NaN   
765   NaN        1.0          NaN          NaN             NaN   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                 0.0                   0.0       2.0          1.0   
1                 0.0                   0.0       0.0          2.0   
2                 1.0                   4.0       2.0          2.0   
3                 0.0                   2.0       2.0          2.0   
4                 1.0                   2.0       1.0          1.0   
..                ...                   ...       ...          ...   
761               NaN                   NaN       NaN          NaN   
762               1.0                   4.0       0.0          1.0   
763               NaN                   NaN       NaN          NaN   
764               NaN                   NaN       NaN          NaN   
765               NaN                   NaN       NaN          NaN   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                         0.0   
1                          4.0  ...                         0.0   
2                          3.0  ...                         0.0   
3                          0.0  ...                         0.0   
4                          0.5  ...                         0.0   
..                         ...  ...                         ...   
761                        NaN  ...                         1.0   
762                        6.0  ...                         0.0   
763                        5.2  ...                         0.0   
764                        NaN  ...                         NaN   
765                        NaN  ...                         NaN   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0           0.0           1.0        1.0         1.0   
1                     1.0           1.0           1.0        1.0         0.0   
2                     0.0           1.0           1.0        0.0         0.0   
3                     0.0           0.0           1.0        1.0         0.0   
4                     0.0           0.0           1.0        1.0         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0           0.0           1.0        1.0         NaN   
762                   0.0           1.0           0.0        1.0         0.0   
763                   0.0           1.0           1.0        1.0         0.0   
764                   NaN           NaN           NaN        NaN         NaN   
765                   NaN           NaN           NaN        NaN         NaN   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                        0.0                        0.0   
1           0.0                        0.0                        0.0   
2           NaN                        0.0                        NaN   
3           0.0                        0.0                        0.0   
4           0.0                        0.0                        0.0   
..          ...                        ...                        ...   
761         0.0                        NaN                        0.0   
762         0.0       

<b>- replace NaN values with mean of each column -</b>

In [61]:
df_mean = df.fillna(df.mean())

In [62]:
df_mean

age  care_team  education_3  education_2  marital_status  \
0    66.000000        0.0     1.000000     0.000000        4.000000   
1    54.000000        0.0     2.000000     1.000000        4.000000   
2    45.000000        0.0     2.000000     1.000000        0.000000   
3    70.000000        0.0     1.000000     0.000000        4.000000   
4    49.000000        0.0     3.000000     1.000000        1.000000   
..         ...        ...          ...          ...             ...   
761  54.108322        3.0     2.462667     0.921333        1.328859   
762  43.000000        3.0     3.000000     1.000000        0.000000   
763  54.108322        3.0     2.462667     0.921333        1.328859   
764  54.108322        0.0     2.462667     0.921333        1.328859   
765  54.108322        1.0     2.462667     0.921333        1.328859   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0             0.00000               0.00000  2.000000     1.000000   
1             0.00000               0.00000  0.000000     2.000000   
2             1.00000               4.00000  2.000000     2.000000   
3             0.00000               2.00000  2.000000     2.000000   
4             1.00000               2.00000  1.000000     1.000000   
..                ...                   ...       ...          ...   
761           0.85906               1.32672  0.630986     1.034713   
762           1.00000               4.00000  0.000000     1.000000   
763           0.85906               1.32672  0.630986     1.034713   
764           0.85906               1.32672  0.630986     1.034713   
765           0.85906               1.32672  0.630986     1.034713   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                     0.000000  ...                    0.000000   
1                     4.000000  ...                    0.000000   
2                     3.000000  ...                    0.000000   
3                     0.000000  ...                    0.000000   
4                     0.500000  ...                    0.000000   
..                         ...  ...                         ...   
761                   4.218319  ...                    1.000000   
762                   6.000000  ...                    0.000000   
763                   5.200000  ...                    0.000000   
764                   4.218319  ...                    0.296149   
765                   4.218319  ...                    0.296149   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                0.000000      0.000000       1.00000   1.000000    1.000000   
1                1.000000      1.000000       1.00000   1.000000    0.000000   
2                0.000000      1.000000       1.00000   0.000000    0.000000   
3                0.000000      0.000000       1.00000   1.000000    0.000000   
4                0.000000      0.000000       1.00000   1.000000    0.000000   
..                    ...           ...           ...        ...         ...   
761              0.000000      0.000000       1.00000   1.000000    0.264228   
762              0.000000      1.000000       0.00000   1.000000    0.000000   
763              0.000000      1.000000       1.00000   1.000000    0.000000   
764              0.083558      0.501992       0.74574   0.816514    0.264228   
765              0.083558      0.501992       0.74574   0.816514    0.264228   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0       0.00000                   0.000000                   0.000000   
1       0.00000                   0.000000                   0.000000   
2       0.00768                   0.000000                   0.348765   
3       0.00000                   0.000000                   0.000000   
4       0.00000                   0.000000                   0.000000   
..          ...                        ...                        ...   
761     0.00000                   0.2829

<b>- roundup values</b> (<i>age</i>, <i>education_3</i>, <i>education_2</i>, <i>marital_status</i>,<i>marital_status_2</i>, <i>M0_employment_status</i>, <i>income_3</i>, <i>M0_exercise</i>,<i>nccn_distress_thermometer</i>, <i>m3_nccn_distress_thermometer</i>, <i>M0_optimism_LOT</i>, <i>M0_mental_health_total_HADS</i>,<i>M3_emotional_support</i>, <i>M3_social_support_total</i>, <i>M3_PTGI_spiritual_change</i>, <i>M3_mental_health_total_HADS</i>,<i>cancer_stage</i>, <i>cancer_grade</i>, <i>baseline_chronic_illness</i>, <i>baseline_histological_type</i>,<i>M0_performancestatus</i>, <i>chemotherapy</i>,<i>radiotherapy</i>,<i>endocrine</i>, <i>M0_surgery</i>,<i>M3_surgery</i>,<i>m0_Global_QLQ30groups65gl</i>,<i>m3_Global_QLQ30groups65gl</i>, <i>m18_Global_QLQ30groups65gl</i>) -

In [63]:
df_mean.age = df_mean.age.round()
df_mean.education_3 = df_mean.education_3.round()
df_mean.education_2 = df_mean.education_2.round()
df_mean.marital_status = df_mean.marital_status.round()
df_mean.marital_status_2 = df_mean.marital_status_2.round()
df_mean.M0_employment_status = df_mean.M0_employment_status.round()
df_mean.income_3 = df_mean.income_3.round()
df_mean.M0_exercise = df_mean.M0_exercise.round()
df_mean.nccn_distress_thermometer = df_mean.nccn_distress_thermometer.round()
df_mean.m3_nccn_distress_thermometer = df_mean.m3_nccn_distress_thermometer.round()
df_mean.M0_optimism_LOT = df_mean.M0_optimism_LOT.round()
df_mean.M0_mental_health_total_HADS = df_mean.M0_mental_health_total_HADS.round()
df_mean.M3_emotional_support = df_mean.M3_emotional_support.round()
df_mean.M3_social_support_total = df_mean.M3_social_support_total.round()
df_mean.M3_PTGI_spiritual_change = df_mean.M3_PTGI_spiritual_change.round()
df_mean.M3_mental_health_total_HADS = df_mean.M3_mental_health_total_HADS.round()
df_mean.cancer_stage = df_mean.cancer_stage.round()
df_mean.baseline_chronic_illness = df_mean.baseline_chronic_illness.round()
df_mean.baseline_histological_type = df_mean.baseline_histological_type.round()
df_mean.M0_performancestatus = df_mean.M0_performancestatus.round()
df_mean.chemotherapy = df_mean.chemotherapy.round()
df_mean.radiotherapy = df_mean.radiotherapy.round()
df_mean.endocrine = df_mean.endocrine.round()
df_mean.M0_surgery = df_mean.M0_surgery.round()
df_mean.M3_surgery = df_mean.M3_surgery.round()
df_mean.m0_Global_QLQ30groups65gl = df_mean.m0_Global_QLQ30groups65gl.round()
df_mean.m3_Global_QLQ30groups65gl = df_mean.m3_Global_QLQ30groups65gl.round()
df_mean.m18_Global_QLQ30groups65gl = df_mean.m18_Global_QLQ30groups65gl.round()

In [64]:
df_mean

age  care_team  education_3  education_2  marital_status  \
0    66.0        0.0          1.0          0.0             4.0   
1    54.0        0.0          2.0          1.0             4.0   
2    45.0        0.0          2.0          1.0             0.0   
3    70.0        0.0          1.0          0.0             4.0   
4    49.0        0.0          3.0          1.0             1.0   
..    ...        ...          ...          ...             ...   
761  54.0        3.0          2.0          1.0             1.0   
762  43.0        3.0          3.0          1.0             0.0   
763  54.0        3.0          2.0          1.0             1.0   
764  54.0        0.0          2.0          1.0             1.0   
765  54.0        1.0          2.0          1.0             1.0   

     marital_status_2  M0_employment_status  income_3  M0_exercise  \
0                 0.0                   0.0       2.0          1.0   
1                 0.0                   0.0       0.0          2.0   
2                 1.0                   4.0       2.0          2.0   
3                 0.0                   2.0       2.0          2.0   
4                 1.0                   2.0       1.0          1.0   
..                ...                   ...       ...          ...   
761               1.0                   1.0       1.0          1.0   
762               1.0                   4.0       0.0          1.0   
763               1.0                   1.0       1.0          1.0   
764               1.0                   1.0       1.0          1.0   
765               1.0                   1.0       1.0          1.0   

     nccn_distress_thermometer  ...  baseline_histological_type  \
0                          0.0  ...                         0.0   
1                          4.0  ...                         0.0   
2                          3.0  ...                         0.0   
3                          0.0  ...                         0.0   
4                          0.0  ...                         0.0   
..                         ...  ...                         ...   
761                        4.0  ...                         1.0   
762                        6.0  ...                         0.0   
763                        5.0  ...                         0.0   
764                        4.0  ...                         0.0   
765                        4.0  ...                         0.0   

     M0_performancestatus  chemotherapy  radiotherapy  endocrine  M0_surgery  \
0                     0.0           0.0           1.0        1.0         1.0   
1                     1.0           1.0           1.0        1.0         0.0   
2                     0.0           1.0           1.0        0.0         0.0   
3                     0.0           0.0           1.0        1.0         0.0   
4                     0.0           0.0           1.0        1.0         0.0   
..                    ...           ...           ...        ...         ...   
761                   0.0           0.0           1.0        1.0         0.0   
762                   0.0           1.0           0.0        1.0         0.0   
763                   0.0           1.0           1.0        1.0         0.0   
764                   0.0           1.0           1.0        1.0         0.0   
765                   0.0           1.0           1.0        1.0         0.0   

     M3_surgery  m0_Global_QLQ30groups65gl  m3_Global_QLQ30groups65gl  \
0           0.0                        0.0                        0.0   
1           0.0                        0.0                        0.0   
2           0.0                        0.0                        0.0   
3           0.0                        0.0                        0.0   
4           0.0                        0.0                        0.0   
..          ...                        ...                        ...   
761         0.0                        0.0                        0.0   
762         0.0       

--------------

<font color='blue'><b><u>SVM - Support Vector Machine (ten fold)</u></b></font>

In [71]:
from sklearn.svm import SVC
from sklearn.model_selection import KFold
from sklearn.model_selection import cross_val_score
from sklearn.model_selection import cross_val_predict

In [72]:
# split the dataset into independent (X) and dependent (Y) (m12) which is the target value
X=df_mean.iloc[:,1:-1].values
Y=df_mean.iloc[:,-1].values

In [73]:
X_train, X_test, Y_train, Y_test = X[train_index], X[test_index], Y[train_index], Y[test_index]

In [74]:
classifier = SVC(kernel='rbf')

In [75]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train,Y_train)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.7236842105263158


<font color='red'>Accuracy: 72.37%</font>

In [76]:
y_predict = classifier.predict(X_test)

In [77]:
#confusion matrix
pd.crosstab(Y_test,y_predict)

col_0  0.0
row_0     
0.0     55
1.0     21

For 0: <font color='blue'>100%</font> (55/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>0%</font> (0/21) correctly classified -> <font color='red'>poor QoL</font> accuracy

----------------

<u><font color='blue'>SMOTE</u></font>

In [78]:
from sklearn.model_selection import train_test_split
from imblearn.over_sampling import SMOTE
smote = SMOTE()

In [79]:
X_train_smote, Y_train_smote = smote.fit_resample(X_train.astype('float'),Y_train)

In [80]:
from collections import Counter
print("Before SMOTE: ", Counter(Y_train))
print("After SMOTE: ", Counter(Y_train_smote))

Before SMOTE:  Counter({0.0: 517, 1.0: 173})
After SMOTE:  Counter({0.0: 517, 1.0: 517})


In [81]:
from sklearn.metrics import accuracy_score
classifier.fit(X_train_smote,Y_train_smote)
Y_predict=classifier.predict(X_test)
print('Accuracy:',accuracy_score(Y_test,Y_predict))

Accuracy: 0.75


<font color='red'>Accuracy: 75%</font>

In [82]:
pd.crosstab(Y_test,Y_predict)

col_0  0.0  1.0
row_0          
0.0     42   13
1.0      6   15

For 0: <font color='blue'>76%</font> (42/55) correctly classified - > <font color='blue'>good QoL</font> accuracy
<br>
For 1: <font color='red'>71%</font> (15/21) correctly classified -> <font color='red'>poor QoL</font> accuracy